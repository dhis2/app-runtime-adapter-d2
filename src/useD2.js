import { useState, useEffect } from 'react'
import { init as initD2, config, getUserSettings } from 'd2'
import { useConfig } from '@dhis2/app-runtime'

let theD2 = null

const configI18n = async (baseUrl, i18nRoot) => {
  config.baseUrl = baseUrl

  const settings = await getUserSettings()

  if (settings.keyUiLocale && settings.keyUiLocale !== 'en') {
    config.i18n.sources.add(
      `${i18nRoot}/i18n_module_${settings.keyUiLocale}.properties`
    )
  }

  config.i18n.sources.add('${i18nRoot}/i18n_module_en.properties')
}

const init = async ({ appUrl, baseUrl, d2Config, i18nRoot = null }) => {
  if (i18nRoot) {
    await configI18n(baseUrl, i18nRoot)
  }

  return await initD2({
    appUrl,
    baseUrl,
    ...d2Config,
  })
}

export const useD2 = ({
  d2Config = {},
  onInitialized = Function.prototype,
  i18nRoot,
}) => {
  const { baseUrl, apiVersion } = useConfig()
  const [d2, setD2] = useState(theD2)
  const [d2Error, setError] = useState(undefined)

  useEffect(() => {
    if (!theD2) {
      init({
        appUrl: baseUrl,
        baseUrl: `${baseUrl}/api/${apiVersion}`,
        d2Config,
        i18nRoot
      })
        .then(async (d2) => {
          await onInitialized(d2)
          theD2 = d2
          setD2(d2)
        })
        .catch(setError)
    }
  }, [])

  return { d2, d2Error } // d2 is null while loading
}
