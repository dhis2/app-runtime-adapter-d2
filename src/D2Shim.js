import * as PropTypes from 'prop-types'
import { useD2 } from './useD2'

export const D2Shim = ({ children, onInitialized, d2Config, i18nRoot }) => {
    const { d2, d2Error } = useD2({ onInitialized, d2Config, i18nRoot })

    return children({ d2, d2Error })
}

D2Shim.propTypes = {
    children: PropTypes.func.isRequired,
    d2Config: PropTypes.object,
    i18nRoot: PropTypes.string,
    onInitialized: PropTypes.func,
}
