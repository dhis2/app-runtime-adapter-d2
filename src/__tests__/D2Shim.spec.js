/* eslint-disable no-import-assign,import/namespace */

import React from 'react'
import { D2Shim } from '../D2Shim'
import { render } from '@testing-library/react'
import * as useD2 from '../useD2'
import '@testing-library/jest-dom'

// this polyfill shouldn't be needed once Jest is upgraded to 26
window.MutationObserver = require('mutation-observer')

jest.mock('../useD2')

describe('D2Shim', () => {
    beforeEach(() => {
        jest.restoreAllMocks()
    })
    it('returns a component with children passed in', () => {
        useD2.useD2 = jest
            .fn()
            .mockReturnValue({ d2: null, d2Error: undefined })

        const { container } = render(
            <D2Shim>
                {({ d2 }) => {
                    if (!d2) {
                        return null
                    }
                    return <div d2={d2} />
                }}
            </D2Shim>
        )

        expect(container.firstChild).toMatchSnapshot()
    })
    it('returns a component with children passed in', () => {
        useD2.useD2 = jest
            .fn()
            .mockReturnValue({ d2: 'd2obj', d2Error: undefined })

        const { container } = render(
            <D2Shim>
                {({ d2 }) => {
                    if (!d2) {
                        return null
                    }
                    return <div d2={d2} />
                }}
            </D2Shim>
        )

        expect(container.firstChild).toMatchSnapshot()
    })
})
