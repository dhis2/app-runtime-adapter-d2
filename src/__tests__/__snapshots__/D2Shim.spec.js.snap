// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`D2Shim returns a component with children passed in 1`] = `null`;

exports[`D2Shim returns a component with children passed in 2`] = `
<div
  d2="d2obj"
/>
`;
